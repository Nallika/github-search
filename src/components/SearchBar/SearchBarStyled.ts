import styled from 'styled-components';

const Wrapper = styled.section`
  display: flex;
`;

const SearchInput = styled.input`
  font-size: 1.2em;
  border: 2px solid gray;
  padding: 5px;
  height: 50px;
  width: 100%;
  border-top-left-radius: 15px;
  border-bottom-left-radius: 15px;
  box-sizing: border-box;

  &:focus {
    outline-width: 0;
  }
`;

const SearchButton = styled.button`
  font-size: 1.2em;
  height: 50px;
  width: 80px;
  border-top-right-radius: 15px;
  border-bottom-right-radius: 15px;
`;

export { Wrapper, SearchInput, SearchButton }
