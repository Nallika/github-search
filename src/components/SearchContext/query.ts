import { gql } from '@apollo/client';

export const GET_REPOS = gql`
  query($query: String!, $first: Int!, $after: String) {
    search(query: $query, type: REPOSITORY, first: $first, after: $after) {
      repositoryCount
      pageInfo {
        endCursor
      }
      edges {
        node {
          ... on Repository {
            id
            name
            stargazerCount
            forkCount
            url
          }
        }
      }
    }
  }
`;
