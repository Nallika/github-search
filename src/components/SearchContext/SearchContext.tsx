import React, {createContext, useCallback, useContext, useState} from 'react';
import { useLazyQuery } from '@apollo/client';

import { SearchContextValues, SearchProviderProps } from '../../types';
import { GET_REPOS } from './query';
import { DEFAULT_OFFSET } from '../../constants';

const SearchContext = createContext<SearchContextValues>({
  results: [],
  count: 0,
  loading: false,
  requestSearch: () => {}
});

export const SearchProvider: React.FC<SearchProviderProps> = ({ children }) => {
  const [getData, { loading, data }] = useLazyQuery(GET_REPOS);
  const [ searchQuery, setSearchQuery ] = useState('');
  const [ searchResults, setSearchResult ] = useState([]);
  const [ resultsCount, setResultsCount ] = useState(0);

  const requestSearch = useCallback(async (query: any) => {
    if (loading) {
      return;
    }

    const isNewSearch = query !== searchQuery;

    try {
      const offset = isNewSearch ? null : data.search.pageInfo.endCursor;

      const { data: { search: {edges: results, repositoryCount}} } = await getData({
        variables: {
          query,
          first: DEFAULT_OFFSET,
          after: offset
        }
      });

      setResultsCount(repositoryCount);
      setSearchQuery(query);
      setSearchResult(results);

    } catch (error) {
      console.error(error);
    }
  }, [data, getData, loading, searchQuery]);

  return (
    <SearchContext.Provider value={{
      results: searchResults,
      count: resultsCount,
      loading,
      requestSearch
    }}>
      {children}
    </SearchContext.Provider>
  );
};

export const useSearch = () => useContext(SearchContext);
