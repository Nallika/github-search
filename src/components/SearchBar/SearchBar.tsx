import React, { ChangeEvent, useState } from 'react';

import { Wrapper, SearchInput, SearchButton } from './SearchBarStyled';
import { useSearch } from '../SearchContext';

/**
 * Display input and button for request new search by provided query
 * @constructor
 */
const SearchBar = () => {
  const { requestSearch } = useSearch();
  const [ searchQuery, setSearchQuery ] = useState('');

  const onInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(event.target.value);
  }

  const handleSearch = () => {
    requestSearch(searchQuery);
  }

  return (
    <Wrapper>
      <SearchInput
        data-testid={'search-input'}
        type={'text'}
        onChange={onInputChange}
        placeholder={`Enter search query`}
      />
      <SearchButton data-testid={'search-button'}  onClick={handleSearch} >{'Search'}</SearchButton>
    </Wrapper>
  );
}

export default SearchBar;
