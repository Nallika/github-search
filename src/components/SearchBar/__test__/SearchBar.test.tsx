import React from 'react';
import { screen, render, fireEvent } from '@testing-library/react';
import SearchBar from '../SearchBar';
import * as SearchContext from '../../SearchContext';

jest.mock('../../SearchContext', () => ({
  ...jest.requireActual('../../SearchContext'),
  useSearch: jest.fn(),
}));

describe('SearchBar', () => {
  it('`<SearchBar />` Should render component', () => {
    const mockRequestSearch = jest.fn();
    (SearchContext.useSearch as jest.Mock).mockReturnValue({
      requestSearch: mockRequestSearch,
    });

    const { container } = render(<SearchBar />);

    expect(container).toMatchSnapshot();
  });

  it('calls requestSearch with the entered query on button click', () => {
    const mockRequestSearch = jest.fn();
    (SearchContext.useSearch as jest.Mock).mockReturnValue({
      requestSearch: mockRequestSearch,
    });

    render(<SearchBar />);

    const inputElement = screen.getByTestId('search-input');
    const buttonElement = screen.getByTestId('search-button');

    fireEvent.change(inputElement, { target: { value: 'test query' } });
    fireEvent.click(buttonElement);

    expect(mockRequestSearch).toHaveBeenCalledWith('test query');
  });
});
