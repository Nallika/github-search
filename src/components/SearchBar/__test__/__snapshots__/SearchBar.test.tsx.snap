// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar \`<SearchBar />\` Should render component 1`] = `
<div>
  <section
    class="sc-bdfCDU kQfDmq"
  >
    <input
      class="sc-gsTDqH crDIUr"
      data-testid="search-input"
      placeholder="Enter search query"
      type="text"
    />
    <button
      class="sc-dlfmHC jjLyAb"
      data-testid="search-button"
    >
      Search
    </button>
  </section>
</div>
`;
